File: app/workflow.py
from __future__ import annotations

from datetime import datetime
from typing import Dict, Iterable, List

from app.domain import (
    DebugChangeLog,
    Device,
    DeviceStatus,
    InspectionResult,
    InspectionTask,
    ProgramVersion,
    VersionType,
)


class WorkflowError(ValueError):
    """Raised when business constraints are violated."""


class DeviceWorkflowService:
    """Core workflow service for v1 local validation."""

    def __init__(self) -> None:
        self._devices: Dict[str, Device] = {}

    def create_device(
        self,
        device_sn: str,
        product_line: str,
        model: str,
        owner_debugger: str,
        baseline_file_name: str,
        created_by: str,
    ) -> Device:
        if device_sn in self._devices:
            raise WorkflowError(f"device already exists: {device_sn}")
        now = datetime.utcnow()
        device = Device(
            device_sn=device_sn,
            product_line=product_line,
            model=model,
            owner_debugger=owner_debugger,
            created_at=now,
        )
        device.program_versions.append(
            ProgramVersion(
                version_no="v1.0.0",
                version_type=VersionType.BASELINE,
                file_name=baseline_file_name,
                uploaded_by=created_by,
                uploaded_at=now,
                note="baseline",
            )
        )
        self._devices[device_sn] = device
        return device

    def add_program_version(
        self,
        device_sn: str,
        version_no: str,
        version_type: VersionType,
        file_name: str,
        uploaded_by: str,
        note: str = "",
        sealed: bool = False,
        inspection_id: str | None = None,
    ) -> ProgramVersion:
        device = self.get_device(device_sn)
        if device.status in {DeviceStatus.INSPECTED, DeviceStatus.FROZEN} and version_type != VersionType.FINAL:
            raise WorkflowError("inspected/frozen devices cannot add non-final versions in v1")

        if version_type == VersionType.FINAL and not sealed:
            raise WorkflowError("final version must be sealed")

        version = ProgramVersion(
            version_no=version_no,
            version_type=version_type,
            file_name=file_name,
            uploaded_by=uploaded_by,
            uploaded_at=datetime.utcnow(),
            note=note,
            sealed=sealed,
            inspection_id=inspection_id,
        )
        device.program_versions.append(version)
        self._try_promote_to_inspected(device)
        return version

    def add_debug_log(self, device_sn: str, log: DebugChangeLog) -> None:
        device = self.get_device(device_sn)
        if device.status != DeviceStatus.DEBUGGING:
            raise WorkflowError("debug logs can only be added while debugging")
        device.change_logs.append(log)

    def submit_for_inspection(self, device_sn: str, created_by: str, checklist_codes: Iterable[str]) -> InspectionTask:
        device = self.get_device(device_sn)
        if device.status != DeviceStatus.DEBUGGING:
            raise WorkflowError("only debugging devices can submit inspection")
        if not device.has_baseline():
            raise WorkflowError("baseline version required")
        if device.has_overdue_logs():
            raise WorkflowError("device has overdue (>24h) debug logs")

        task = InspectionTask(
            task_id=f"insp-{device.device_sn}-{int(datetime.utcnow().timestamp())}",
            device_sn=device.device_sn,
            created_by=created_by,
            created_at=datetime.utcnow(),
            results=[InspectionResult(item_code=code, required=True, passed=False) for code in checklist_codes],
        )
        device.inspection_task = task
        device.status = DeviceStatus.PENDING_INSPECTION
        return task

    def sign_inspection(self, device_sn: str, signed_by: str, results: List[InspectionResult]) -> None:
        device = self.get_device(device_sn)
        task = device.inspection_task
        if not task:
            raise WorkflowError("inspection task not found")

        task.results = results
        task.signed_by = signed_by
        task.signed_at = datetime.utcnow()
        self._try_promote_to_inspected(device)

    def mark_documents_ready(self, device_sn: str, ready: bool = True) -> None:
        device = self.get_device(device_sn)
        device.documents_ready = ready
        self._try_promote_to_inspected(device)

    def deliver(self, device_sn: str) -> None:
        device = self.get_device(device_sn)
        if device.status != DeviceStatus.INSPECTED:
            raise WorkflowError("only inspected devices can be delivered")
        device.status = DeviceStatus.DELIVERED

    def freeze(self, device_sn: str) -> None:
        device = self.get_device(device_sn)
        if device.status != DeviceStatus.DELIVERED:
            raise WorkflowError("only delivered devices can be frozen")
        device.status = DeviceStatus.FROZEN

    def get_device(self, device_sn: str) -> Device:
        if device_sn not in self._devices:
            raise WorkflowError(f"device not found: {device_sn}")
        return self._devices[device_sn]

    def _try_promote_to_inspected(self, device: Device) -> None:
        task = device.inspection_task
        if not task:
            return
        if task.all_required_passed() and device.has_final_sealed() and device.documents_ready:
            device.status = DeviceStatus.INSPECTED

File: web/workflow.js
class WorkflowError extends Error {}

const DeviceStatus = {
  DEBUGGING: 'debugging',
  PENDING_INSPECTION: 'pending_inspection',
  INSPECTED: 'inspected',
  DELIVERED: 'delivered',
  FROZEN: 'frozen',
};

const VersionType = {
  BASELINE: 'baseline',
  PROCESS: 'process',
  FINAL: 'final',
};

class DeviceWorkflowService {
  constructor() {
    this.devices = new Map();
    this.frozenTemplates = [
      { templateVersion: 'v1', productLine: 'L1', model: '*', items: ['C-AXIS', 'C-INTERLOCK', 'C-ALARM'] },
      { templateVersion: 'v1', productLine: 'L2', model: '*', items: ['C-SAFETY', 'C-I/O'] },
    ];
  }

  assertRole(role, allow) {
    if (!allow.includes(role)) throw new WorkflowError(`permission denied for role=${role}`);
  }

  createDevice(payload, role, user) {
    this.assertRole(role, ['debugger', 'admin']);
    const { deviceSn, productLine, model, ownerDebugger, baselineFileName } = payload;
    if (this.devices.has(deviceSn)) throw new WorkflowError('device already exists');
    const now = new Date().toISOString();
    const device = {
      deviceSn,
      productLine,
      model,
      ownerDebugger,
      status: DeviceStatus.DEBUGGING,
      documentsReady: false,
      inspectionReportArchived: false,
      techDocsApproved: false,
      createdAt: now,
      programVersions: [{
        versionNo: 'v1.0.0', versionType: VersionType.BASELINE, fileName: baselineFileName,
        uploadedBy: user, uploadedAt: now, note: 'baseline', sealed: false, approved: false,
      }],
      changeLogs: [],
      inspectionTask: null,
    };
    this.devices.set(deviceSn, device);
    return device;
  }

  approveBaseline(deviceSn, role, user, approved = true) {
    this.assertRole(role, ['owner', 'admin']);
    const device = this.getDevice(deviceSn);
    const baseline = device.programVersions.find(v => v.versionType === VersionType.BASELINE);
    if (!baseline) throw new WorkflowError('baseline missing');
    baseline.approved = !!approved;
    baseline.approvedBy = user;
    baseline.approvedAt = new Date().toISOString();
    return device;
  }

  addDebugLog(deviceSn, log, role, user) {
    this.assertRole(role, ['debugger', 'admin']);
    const device = this.getDevice(deviceSn);
    if (device.status !== DeviceStatus.DEBUGGING) throw new WorkflowError('only debugging devices can be modified');
    device.changeLogs.push({ ...log, changedBy: user });
    return device;
  }

  addProgramVersion(deviceSn, data, role, user) {
    this.assertRole(role, ['debugger', 'owner', 'admin']);
    const device = this.getDevice(deviceSn);
    if ((device.status === DeviceStatus.INSPECTED || device.status === DeviceStatus.FROZEN) && data.versionType !== VersionType.FINAL) {
      throw new WorkflowError('cannot add non-final version after inspected/frozen');
    }
    if (data.versionType === VersionType.FINAL && role === 'debugger') {
      throw new WorkflowError('debugger cannot seal final version');
    }
    if (data.versionType === VersionType.FINAL && !data.sealed) {
      throw new WorkflowError('final version must be sealed');
    }
    device.programVersions.push({ ...data, uploadedBy: user, uploadedAt: new Date().toISOString() });
    this.tryPromote(device);
    return device;
  }

  submitInspection(deviceSn, role, user) {
    this.assertRole(role, ['debugger', 'owner', 'admin']);
    const device = this.getDevice(deviceSn);
    if (device.status !== DeviceStatus.DEBUGGING) throw new WorkflowError('invalid status');
    const baseline = device.programVersions.find(v => v.versionType === VersionType.BASELINE);
    if (!baseline) throw new WorkflowError('baseline required');
    if (!baseline.approved) throw new WorkflowError('baseline approval required');
    if (device.changeLogs.some(l => (new Date(l.loggedAt) - new Date(l.changedAt)) / 36e5 > 24)) {
      throw new WorkflowError('overdue logs found');
    }

    const checklistCodes = this.generateChecklistItems(device.productLine, device.model);
    device.inspectionTask = {
      taskId: `insp-${deviceSn}-${Date.now()}`,
      createdBy: user,
      createdAt: new Date().toISOString(),
      signedBy: null,
      signedAt: null,
      templateVersion: this.matchTemplate(device.productLine, device.model).templateVersion,
      results: checklistCodes.map(code => ({ itemCode: code, required: true, passed: false, comment: '' })),
    };
    device.status = DeviceStatus.PENDING_INSPECTION;
    return device;
  }

  signInspection(deviceSn, results, role, user) {
    this.assertRole(role, ['owner', 'admin']);
    const device = this.getDevice(deviceSn);
    if (!device.inspectionTask) throw new WorkflowError('inspection task missing');
    device.inspectionTask.results = results;
    device.inspectionTask.signedBy = user;
    device.inspectionTask.signedAt = new Date().toISOString();
    this.tryPromote(device);
    return device;
  }

  setDocumentsReady(deviceSn, ready, role) {
    this.assertRole(role, ['owner', 'admin']);
    const device = this.getDevice(deviceSn);
    device.documentsReady = !!ready;
    this.tryPromote(device);
    return device;
  }

  archiveInspectionReport(deviceSn, role) {
    this.assertRole(role, ['owner', 'admin']);
    const device = this.getDevice(deviceSn);
    device.inspectionReportArchived = true;
    this.tryPromote(device);
    return device;
  }

  approveTechDocs(deviceSn, role) {
    this.assertRole(role, ['owner', 'admin']);
    const device = this.getDevice(deviceSn);
    device.techDocsApproved = true;
    this.tryPromote(device);
    return device;
  }

  deliver(deviceSn, role) {
    this.assertRole(role, ['owner', 'admin']);
    const device = this.getDevice(deviceSn);
    if (device.status !== DeviceStatus.INSPECTED) throw new WorkflowError('must be inspected');
    device.status = DeviceStatus.DELIVERED;
    return device;
  }

  freeze(deviceSn, role) {
    this.assertRole(role, ['owner', 'admin']);
    const device = this.getDevice(deviceSn);
    if (device.status !== DeviceStatus.DELIVERED) throw new WorkflowError('must be delivered');
    device.status = DeviceStatus.FROZEN;
    return device;
  }

  listDevices(role, user) {
    const all = [...this.devices.values()];
    if (role === 'debugger') return all.filter(d => d.ownerDebugger === user);
    return all;
  }

  getDevice(deviceSn) {
    const d = this.devices.get(deviceSn);
    if (!d) throw new WorkflowError('device not found');
    return d;
  }

  matchTemplate(productLine, model) {
    return this.frozenTemplates.find(t => t.productLine === productLine && (t.model === model || t.model === '*'))
      || this.frozenTemplates.find(t => t.productLine === productLine)
      || this.frozenTemplates[0];
  }

  generateChecklistItems(productLine, model) {
    return [...this.matchTemplate(productLine, model).items];
  }

  tryPromote(device) {
    const task = device.inspectionTask;
    if (!task) return;
    const required = task.results.filter(r => r.required);
    const passed = required.length > 0 && required.every(r => r.passed);
    const finalSealed = device.programVersions.some(v => v.versionType === VersionType.FINAL && v.sealed);
    if (passed && finalSealed && device.documentsReady && device.inspectionReportArchived && device.techDocsApproved) {
      device.status = DeviceStatus.INSPECTED;
    }
  }
}

module.exports = { DeviceWorkflowService, WorkflowError, DeviceStatus, VersionType };
